File: android/app/src/main/java/com/anthra/timer/activity/StepCounterNormalizer.kt
package com.anthra.timer.activity

import kotlin.math.max

data class StepCounterState(
  val dayKey: String,
  val timezone: String,
  val bootCount: Int,
  val baselineRaw: Long,
  val lastRaw: Long,
  val steps: Long
)

data class StepCounterUpdate(
  val state: StepCounterState,
  val rolledOverDayKey: String? = null,
  val rolledOverTimezone: String? = null,
  val rolledOverSteps: Long? = null,
  val counterReset: Boolean = false,
  val rebootDetected: Boolean = false,
  val timezoneChanged: Boolean = false
)

object StepCounterNormalizer {
  fun update(
    previous: StepCounterState?,
    rawReading: Long,
    dayKey: String,
    timezone: String,
    bootCount: Int,
    permissionGranted: Boolean
  ): StepCounterUpdate? {
    if (!permissionGranted || rawReading < 0L) return null

    if (previous == null) {
      return StepCounterUpdate(
        state = StepCounterState(dayKey, timezone, bootCount, rawReading, rawReading, 0L)
      )
    }

    val timezoneChanged = previous.timezone != timezone
    val dateChanged = previous.dayKey != dayKey

    if (
      previous.bootCount == bootCount &&
      (
        rawReading < previous.lastRaw ||
          (!timezoneChanged && dayKey < previous.dayKey)
      )
    ) {
      // Reject stale batched callbacks before evaluating their date. Otherwise
      // a late event from yesterday can roll the state backwards into yesterday.
      return StepCounterUpdate(
        state = previous,
        counterReset = true
      )
    }

    if (timezoneChanged && !dateChanged) {
      val delta = if (previous.bootCount == bootCount) {
        rawReading - previous.lastRaw
      } else {
        rawReading
      }
      return StepCounterUpdate(
        state = previous.copy(
          timezone = timezone,
          bootCount = bootCount,
          lastRaw = rawReading,
          steps = max(0L, previous.steps) + max(0L, delta)
        ),
        rebootDetected = previous.bootCount != bootCount,
        timezoneChanged = true
      )
    }
    if (dateChanged) {
      // Events are dated using SensorEvent.timestamp by the manager. With the
      // service continuously subscribed, the first delta observed on a new
      // date belongs to the new date. After a reboot, retaining the new boot's
      // raw count is preferable to silently discarding every post-boot step.
      val firstNewDaySteps = if (previous.bootCount == bootCount) {
        rawReading - previous.lastRaw
      } else {
        rawReading
      }
      return StepCounterUpdate(
        state = StepCounterState(
          dayKey,
          timezone,
          bootCount,
          rawReading - firstNewDaySteps,
          rawReading,
          firstNewDaySteps
        ),
        rolledOverDayKey = previous.dayKey,
        rolledOverTimezone = previous.timezone,
        rolledOverSteps = max(0L, previous.steps),
        rebootDetected = previous.bootCount != bootCount,
        timezoneChanged = timezoneChanged
      )
    }

    if (previous.bootCount != bootCount) {
      // The counter restarted at boot. Its current value represents every step
      // since that reboot, including steps taken while Anthra was not running.
      val nextSteps = max(0L, previous.steps) + rawReading
      return StepCounterUpdate(
        state = StepCounterState(dayKey, timezone, bootCount, rawReading, rawReading, nextSteps),
        rebootDetected = true
      )
    }

    val delta = rawReading - previous.lastRaw
    return StepCounterUpdate(
      state = previous.copy(
        lastRaw = rawReading,
        steps = max(0L, previous.steps) + delta
      )
    )
  }
}

File: modules/anthra-activity/ios/StepCounterNormalizer.swift
import Foundation

struct StepCounterState {
  let dayKey: String
  let timezone: String
  let bootCount: Int
  let baselineRaw: Int64
  let lastRaw: Int64
  let steps: Int64
}

struct StepCounterUpdate {
  let state: StepCounterState
  let rolledOverDayKey: String?
  let rolledOverTimezone: String?
  let rolledOverSteps: Int64?
  let counterReset: Bool
  let rebootDetected: Bool
  let timezoneChanged: Bool
}

enum StepCounterNormalizer {
  static func update(
    previous: StepCounterState?,
    rawReading: Int64,
    dayKey: String,
    timezone: String,
    bootCount: Int,
    permissionGranted: Bool
  ) -> StepCounterUpdate? {
    guard permissionGranted, rawReading >= 0 else { return nil }

    if previous == nil {
      return StepCounterUpdate(
        state: StepCounterState(
          dayKey: dayKey,
          timezone: timezone,
          bootCount: bootCount,
          baselineRaw: rawReading,
          lastRaw: rawReading,
          steps: 0
        ),
        rolledOverDayKey: nil,
        rolledOverTimezone: nil,
        rolledOverSteps: nil,
        counterReset: false,
        rebootDetected: false,
        timezoneChanged: false
      )
    }

    guard let previous else { return nil }
    let timezoneChanged = previous.timezone != timezone
    let dateChanged = previous.dayKey != dayKey

    if previous.bootCount == bootCount
      && (rawReading < previous.lastRaw || (!timezoneChanged && dayKey < previous.dayKey)) {
      return StepCounterUpdate(
        state: previous,
        rolledOverDayKey: nil,
        rolledOverTimezone: nil,
        rolledOverSteps: nil,
        counterReset: true,
        rebootDetected: false,
        timezoneChanged: false
      )
    }

    if timezoneChanged && !dateChanged {
      let delta = previous.bootCount == bootCount ? rawReading - previous.lastRaw : rawReading
      return StepCounterUpdate(
        state: StepCounterState(
          dayKey: previous.dayKey,
          timezone: timezone,
          bootCount: bootCount,
          baselineRaw: previous.baselineRaw,
          lastRaw: rawReading,
          steps: max(0, previous.steps) + max(0, delta)
        ),
        rolledOverDayKey: nil,
        rolledOverTimezone: nil,
        rolledOverSteps: nil,
        counterReset: false,
        rebootDetected: previous.bootCount != bootCount,
        timezoneChanged: true
      )
    }

    if dateChanged {
      let firstNewDaySteps = previous.bootCount == bootCount ? rawReading - previous.lastRaw : rawReading
      return StepCounterUpdate(
        state: StepCounterState(
          dayKey: dayKey,
          timezone: timezone,
          bootCount: bootCount,
          baselineRaw: rawReading - firstNewDaySteps,
          lastRaw: rawReading,
          steps: firstNewDaySteps
        ),
        rolledOverDayKey: previous.dayKey,
        rolledOverTimezone: previous.timezone,
        rolledOverSteps: max(0, previous.steps),
        counterReset: false,
        rebootDetected: previous.bootCount != bootCount,
        timezoneChanged: timezoneChanged
      )
    }

    if previous.bootCount != bootCount {
      let nextSteps = max(0, previous.steps) + rawReading
      return StepCounterUpdate(
        state: StepCounterState(
          dayKey: dayKey,
          timezone: timezone,
          bootCount: bootCount,
          baselineRaw: rawReading,
          lastRaw: rawReading,
          steps: nextSteps
        ),
        rolledOverDayKey: nil,
        rolledOverTimezone: nil,
        rolledOverSteps: nil,
        counterReset: false,
        rebootDetected: true,
        timezoneChanged: false
      )
    }

    let delta = rawReading - previous.lastRaw
    return StepCounterUpdate(
      state: StepCounterState(
        dayKey: dayKey,
        timezone: timezone,
        bootCount: bootCount,
        baselineRaw: previous.baselineRaw,
        lastRaw: rawReading,
        steps: max(0, previous.steps) + delta
      ),
      rolledOverDayKey: nil,
      rolledOverTimezone: nil,
      rolledOverSteps: nil,
      counterReset: false,
      rebootDetected: false,
      timezoneChanged: false
    )
  }
}

struct StepDaySnapshot {
  let dateKey: String
  let timezone: String
  let steps: Int64
}
